packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart

```diff
@@ -0,0 +1,542 @@
+        throw StateError(
+          'Authentication was required but no authentication info could be retrieved.',
+        );
+      }
+
+      void localRevokedAuthenticationCallback(event) async {
+        var authenticationRevokedReason = switch (event) {
+          RevokedAuthenticationUser _ =>
+            AuthenticationFailureReason.unauthenticated,
+          RevokedAuthenticationAuthId revokedAuthId =>
+            revokedAuthId.authId == authenticationInfo.authId
+                ? AuthenticationFailureReason.unauthenticated
+                : null,
+          RevokedAuthenticationScope revokedScopes => revokedScopes.scopes.any(
+              (s) => endpoint.requiredScopes.map((s) => s.name).contains(s),
+            )
+                ? AuthenticationFailureReason.insufficientAccess
+                : null,
+          _ => null,
+        };
+
+        if (authenticationRevokedReason != null) {
+          onRevokedAuthenticationCallback();
+        }
+      }
+
+      session.messages.addListener(
+        MessageCentralServerpodChannels.revokedAuthentication(
+          authenticationInfo.userId,
+        ),
+        localRevokedAuthenticationCallback,
+      );
+
+      _revokedAuthenticationCallback = localRevokedAuthenticationCallback;
+    }
+  }
+}
+
+class _InputStreamContext implements _StreamContext {
+  @override
+  final StreamController controller;
+
+  _InputStreamContext(this.controller);
+}
+
+class _OutputStreamContext implements _StreamContext {
+  @override
+  final StreamController controller;
+
+  final StreamSubscription subscription;
+
+  CloseReason? closeReason;
+
+  _OutputStreamContext(this.controller, this.subscription);
+}
+
+abstract interface class _StreamContext {
+  StreamController get controller;
+}
+
+/// Manages the streams for an endpoint method call.
+/// Should only be used by Serverpod packages.
+@internal
+class MethodStreamManager {
+  static const Duration _closeTimeout = Duration(seconds: 6);
+  final Map<String, _InputStreamContext> _inputStreamContexts = {};
+  final Map<String, _OutputStreamContext> _outputStreamContexts = {};
+
+  MethodStreamManager({
+    this.onInputStreamClosed,
+    this.onOutputStreamClosed,
+    this.onOutputStreamError,
+    this.onOutputStreamValue,
+    this.onAllStreamsClosed,
+  });
+
+  void Function(
+    UuidValue namespace,
+    Object? value,
+    MethodStreamCallContext methodStreamCallContext,
+  )? onOutputStreamValue;
+  void Function(
+    UuidValue namespace,
+    Object error,
+    StackTrace stackTrace,
+    MethodStreamCallContext methodStreamCallContext,
+  )? onOutputStreamError;
+  void Function(
+    UuidValue namespace,
+    CloseReason? closeReason,
+    MethodStreamCallContext callContext,
+  )? onOutputStreamClosed;
+  void Function(
+    UuidValue namespace,
+    String parameterName,
+    CloseReason? closeReason,
+    MethodStreamCallContext callContext,
+  )? onInputStreamClosed;
+
+  void Function()? onAllStreamsClosed;
```

Comment:
I can move them up! 👍  Not sure typedefs are a good idea here since you would only use it once and then you have to jump to the typedefs to see the signatures.

Reply:
[0cb862a](https://github.com/serverpod/serverpod/pull/2763/commits/0cb862ae1412891b6c8b8e4f81bf8b21cc54816c)

Reply:
I think the typedefs would help with the duplication of the function signature in both the field and the named parameter.

Would that be worth the consideration?

Reply:
Ah, I should have re-evaluated after I made the hooks private, now I think it makes sense since there are two places 👍

Reply:
98cd6294